Main review mods 2023 12 06

tlhumphrey2 opened the pull request.

dcamper commented on hpcc/sasha.auto.tfvars.example:
Please remove all empty files from the repo.

tlhumphrey2 replied:
I don't see any empty files. Nothing to do.

dcamper commented on lite-variables.tf:
```diff
@@ -0,0 +1,194 @@
+}
+
+variable "aks_node_sizes" {
+  description = "REQUIRED.  The VM size for each node in the HPCC Systems node pool.\nRecommend \"Standard_B4ms\" or better.\nSee https://docs.microsoft.com/en-us/azure/virtual-machines/sizes-general for more information."
```
This description does not match the value type; please update.

Please verify all descriptions in variable files align with their value types and default values.

tlhumphrey2 replied:
TLH: Fixed

tlhumphrey2 replied:
aks_node_sizes is on longer an input variable. It is now a local variable.

dcamper commented on lite-variables.tf:
```diff
@@ -0,0 +1,194 @@
+  default     = ""
+}
+
+variable "admin_username" {
```
Curious: Why is there a separate entry for HPCC admin and AKS admin?

dcamper replied:
So admin_name and aks_admin_name are really the same but you're asking the user to explicitly set them both?  Can you set one equal to the value of the other so the user explicitly defines only one?  Then you can remove documentation for the one the user doesn't see.

tlhumphrey2 replied:
TLH: 'admin_name' doesn't exist any more (this is its name in your terraform). It is now called 'aks_admin_name'. Here is an example of how it would be set 'aks_admin_name="Timothy L Humphrey". I think you are confusing 'admin_username' with 'admin_name'. An example of setting 'admin_username' is as follows: 'admin_username="tlhumphrey2"

dcamper replied:
Argh!  You are right!  I confused those two things.

dcamper commented on scripts/deploy:
```diff
@@ -0,0 +1,122 @@
+  mkdir $HOME/tflogs
+fi
+
+plan=`$thisdir/mkplan deploy_${name}.plan`
```
The use of backticks in bash scripts has been deprecated for some time now, in favor of using $() syntax.  See https://stackoverflow.com/questions/4708549/what-is-the-difference-between-command-and-command-in-shell-programming (and linked articles within that SO page) for more information.

Please use $() syntax here and in all other bash scripts.

tlhumphrey2 replied:
TLH: Replaced `...` with $(...)

tlhumphrey2 replied:
fixed

dcamper commented on storage/main.tf:
```diff
@@ -1,9 +1,9 @@
 module "storage" {
-  source = "github.com/gfortil/terraform-azurerm-hpcc-storage?ref=HPCC-27615"
+  source = "git@github.com:hpccsystems-solutions-lab/terraform-azurerm-hpcc-storage.git?ref=HPCC-27615-add-rm-0000-cidr"
```
This is citing a specific git branch.  Is that intended?

dcamper reviewed (changes requested):
First round of review.  I will pass this back to you for resolving these items before continuing with an actual flow-of-execution review.

dcamper commented on README.md:
```diff
@@ -1,18 +1,133 @@
+1. Issue `terraform init` to initialize the Terraform modules.
+1. Decide how you want to supply option values to the module during invocation.  There are three possibilities:
+	1. Invoke the `terraform apply` command and enter values for each option as Terraform prompts for it, then enter `yes` at the final prompt to begin building the cluster.
+	1. **Recommended:**  Create a `lite.auto.tfvars` file containing the values for each option, invoke `terraform apply`, then enter `yes` at the final prompt to begin building the cluster.  The easiest way to creat `lite.auto.tfvars` is to copy the example file, `lite.auto.tfvars.example`, and then edit the copy:
```
Minor: "creat" -> "create"

tlhumphrey2 replied:
fixed

dcamper commented:
FYI @BaharF

tlhumphrey2 commented:
TLH: Fixed. Changed 'creat' to 'create'.

From: Dan S. Camper ***@***.***>
Sent: Wednesday, December 6, 2023 7:52 AM
To: hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite ***@***.***>
Cc: Humphrey, Timothy L (RIS-HBE) ***@***.***>; Author ***@***.***>
Subject: Re: [hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite] Main review mods 2023 12 06 (PR #55)


*** External email: use caution ***



@dcamper commented on this pull request.

________________________________

In README.md<https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite/pull/55#discussion_r1417455199>:

> +

+* <font color="red">**Azure CLI**</font> To work with Azure, you will need to install the Azure Command Line tools.  Instructions can be found at [https://docs.microsoft.com/en-us/cli/azure/install-azure-cli](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli).  Even if you think you won't be working with Azure, this module does leverage the command line tools to manipulate network security groups within kubernetes clusters.  TL;DR: Make sure you have the command line tools installed.

+

+* To successfully create everything you will need to have Azure's `Contributor` role plus access to `Microsoft.Authorization/*/Write` and `Microsoft.Authorization/*/Delete` permissions on your subscription.  You may have to create a custom role for this.  Of course, Azure's `Owner` role includes everything so if you're the subscription's owner then you're good to go.

+

+## Installing/Using This Module

+

+1. If necessary, login to Azure.

+       * From the command line, this is usually accomplished with the `az login` command.

+1. Clone this repo to your local system and change current directory.

+       * `git clone -b https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite.git`<https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite.git%60>

+       * `cd terraform-azurerm-hpcc-lite`

+1. Issue `terraform init` to initialize the Terraform modules.

+1. Decide how you want to supply option values to the module during invocation.  There are three possibilities:

+       1. Invoke the `terraform apply` command and enter values for each option as Terraform prompts for it, then enter `yes` at the final prompt to begin building the cluster.

+       1. **Recommended:**  Create a `lite.auto.tfvars` file containing the values for each option, invoke `terraform apply`, then enter `yes` at the final prompt to begin building the cluster.  The easiest way to creat `lite.auto.tfvars` is to copy the example file, `lite.auto.tfvars.example`, and then edit the copy:

Minor: "creat" -> "create"

-
Reply to this email directly, view it on GitHub<https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite/pull/55#pullrequestreview-1767784127>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAIUZYHVYPJ42ACPKZGKT33YICBCBAVCNFSM6AAAAABAJMATMSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONRXG44DIMJSG4>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

________________________________
The information contained in this e-mail message is intended only for the personal and confidential use of the recipient(s) named above. This message may be an attorney-client communication and/or work product and as such is privileged and confidential. If the reader of this message is not the intended recipient or an agent responsible for delivering it to the intended recipient, you are hereby notified that you have received this document in error and that any review, dissemination, distribution, or copying of this message is strictly prohibited. If you have received this communication in error, please notify us immediately by e-mail, and delete the original message.

tlhumphrey2 commented:
TLH: Nothing to do in this email.

From: Dan S. Camper ***@***.***>
Sent: Wednesday, December 6, 2023 7:55 AM
To: hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite ***@***.***>
Cc: Humphrey, Timothy L (RIS-HBE) ***@***.***>; Author ***@***.***>
Subject: Re: [hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite] Main review mods 2023 12 06 (PR #55)


*** External email: use caution ***



FYI @BaharF<https://github.com/BaharF>

-
Reply to this email directly, view it on GitHub<https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite/pull/55#issuecomment-1843052506>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAIUZYFR6MSBA7XYVQLBY4LYICBKZAVCNFSM6AAAAABAJMATMSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNBTGA2TENJQGY>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

________________________________
The information contained in this e-mail message is intended only for the personal and confidential use of the recipient(s) named above. This message may be an attorney-client communication and/or work product and as such is privileged and confidential. If the reader of this message is not the intended recipient or an agent responsible for delivering it to the intended recipient, you are hereby notified that you have received this document in error and that any review, dissemination, distribution, or copying of this message is strictly prohibited. If you have received this communication in error, please notify us immediately by e-mail, and delete the original message.

tlhumphrey2 commented on README.md:
```diff
@@ -1,18 +1,133 @@
+
+* <font color="red">**Azure CLI**</font> To work with Azure, you will need to install the Azure Command Line tools.  Instructions can be found at [https://docs.microsoft.com/en-us/cli/azure/install-azure-cli](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli).  Even if you think you won't be working with Azure, this module does leverage the command line tools to manipulate network secu ...
+
+* To successfully create everything you will need to have Azure's `Contributor` role plus access to `Microsoft.Authorization/*/Write` and `Microsoft.Authorization/*/Delete` permissions on your subscription.  You may have to create a custom role for this.  Of course, Azure's `Owner` role includes everything so if you're the subscription's owner then you're good to go.
```
TLH: These requirements have been verified and are required.

dcamper replied:
Can you validate this through testing, please?  Ideally, have an Azure user who does not have the additional permissions try to create a cluster using this TF code.

I am asking for additional testing because:  The additional permissions were an outcome of something I did when I was writing a version of this code.  That is not a requirement from any of the upstream code repositories, so it should not be a requirement in your derived code.

tlhumphrey2 replied:
I tried to deploy lite from an azure VM in a free trial personal azure account. I couldn't do it because a free trial azure account is not allowed to change quotas, and the terraform needs more vCPUs that the quota. 
Also, I mentioned that the terraform requires my azure VM to have EncryptionAtHost enabled, which I documented, under 'Requirements' in README.md. I'm going to try lite deployment in my virtualbox VM to see if the terraform requires something like EncryptionAtHost.

dcamper replied:
How many vCPUs are available to a free trial account?  What size HPCC cluster did you try to create?  Is it possible to configure a tiny HPCC cluster that will work in a free trial account?

It seems to me that someone wanting to "kick the tires" will use a trial account, unless they are part of a business that already uses Azure, so finding a way to make this work in a constrained resource space seems important.

tlhumphrey2 replied:
Have attempted to valided using a free trial azure personal account. But, aks fails because I need higher quota of VMs for eastus. But, free trial accounts won't let you change quotas.

dcamper replied:
I have temporarily removed the special role you had on your Azure user account that was required for my old TF modules.  Please test your TF code with your regular business account, which now has only a simple Contributor role, and see if the TF deploys correctly.  This will indicate whether any other user needs special permissions or not.

tlhumphrey2 replied:
Yes the special role is needed. Now, I get this error when I attempt to deploy aks:
Error: authorization.RoleAssignmentsClient#Create: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'HumphrTL@risk.regn.net' with object id '35cbdc79-7ef5-4d2c-9b59-61ec21d76aa9' does not have authorization to perform action 'Microsoft.Authorization/roleAssignments/write' over scope '/subscriptions/49219efc-701f-4c7e-a2ac-c600308a69e3/resourceGroups/app-tlhhpccplatform-sandbox-eastus2-19543/providers/Microsoft.Network/virtualNetworks/tlhhpcc-dev-eastus2-vnet/subnets/aks-hpcc-private/providers/Microsoft.Authorization/roleAssignments/280f0607-8f28-0a1c-63ea-a3735dfa0578' or the scope is invalid. If access was recently granted, please refresh your credentials."
 
   with module.aks.module.cluster.azurerm_role_assignment.network_contributor_network,
   on .terraform/modules/aks/modules/cluster/main.tf line 8, in resource "azurerm_role_assignment" "network_contributor_network":
    8: resource "azurerm_role_assignment" "network_contributor_network" {

tlhumphrey2 replied:
I have spoke with Godji and Milind. Godji said that he has the role of Owner and has never tried a role with fewer permissions than Owner. Milind says he has always used TFE to deploy the RBA hpcc terraform, but, believes if you use the terraform outside of TFE you would need elevated permissions.

tlhumphrey2 commented on README.md:
```diff
@@ -1,18 +1,133 @@
+	1. Use -var arguments on the command line when executing the Terraform tool to set each of the values found in the .tfvars file.  This method is useful if you are driving the creation of the cluster from a script.
+1. After the Kubernetes cluster is deployed, your local `kubectl` tool can be used to interact with it.  At some point during the deployment `kubectl` will acquire the login credentials for the cluster and it will be the current context (so any `kubectl` commands you enter will be directed to that cluster by default).
+
+At the end of a successful deployment these items are output:
```
TLH: All outputs shown in the README are output after a successful terraform applyl

dcamper replied:
There are four `output.tf` files. Is the output from the hpcc module the only one executed?

dcamper replied:
Maybe I phrased this wrong. There are four `output.tf` files in your repo:

- vnet/outputs.tf
- storage/outputs.tf
- hpcc/outputs.tf
- aks/outputs.tf

Are you saying that only the hpcc/outputs.tf actually produces visible output?

tlhumphrey2 replied:
fixed

tlhumphrey2 commented on README.md:
```diff
@@ -1,18 +1,133 @@
+
+The following options should be set in your `lite.auto.tfvars` file (or entered interactively, if you choose to not create a file).  Only a few of them have default values. The rest are required.  The 'Updateable' column indicates whether, for any given option, it is possible to successfully apply the update against an already-running HPCC k8s cluster.
+
+|Option|Type|Description|Updatable|
```
TLH: I double-checks and comfirmed that all avaiable options (input variables) are documented and all documented options are being used by the terraform.

tlhumphrey2 commented on aks/locals.tf:
```diff
@@ -1,23 +1,123 @@
+      node_size         = var.aks_node_sizes.thor
+      single_group      = false
+      min_capacity      = 3
+      max_capacity      = 6
```
TLH: Added new variable, 'aks_capacity' which defines the minimum and maximum number of nodes in each node pool.

tlhumphrey2 replied:
TLH: This one will take a while to figure out because of the many different node/VM sizes. The following link lists all the node/VM sizes possibly: https://github.com/hpccsystems-solutions-lab/tlh-oss-terraform-azurerm-aks/blob/make-logging-and-monitoring-optional/modules/node-groups/modules/node-group/local.tf

dcamper replied:
Agreed.  I suggest you make an Executive Decision and support only a few (3?) different sizes.  You can fit requirements into VM sizes the best way you can.

Note that the user does not need to know about your decision.

tlhumphrey2 replied:
TLH: I will use that approach. Thanks.

tlhumphrey2 replied:
Fixed. Now thorpool capacity is set based on the user's values for 'thor_num_workers', 'thor_max_jobs', 'thor_worker_cpus', and thorpool node size selected.

tlhumphrey2 replied:
Continuation of my last reply. Also, the above mentioned calculations also sets the number of cpus and ram each thor worker should have.

tlhumphrey2 commented on hpcc/scripts/delete_ephemeral_storage_accounts:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
TLH: When one asked for external storage accounts to be created, the 4 storage accounts that are created, by default, in the aks resource group should not be created (I assumed these are ephemeral). But 2 are created. This script deletes them IF you have asked for external storage accounts to be created.

tlhumphrey2 replied:
TLH: By default the HPCC module creates 4 storage accounts in the aks resource group. I have assumed these are epemeral storage accounts (mainly because Godji created the 'storage' module to create external storage which I assumed replaces the 4 storage accounts created in the aks resource group by default). These 4 storage accounts are used by just HPCC.

dcamper replied:
_A reference to the offending code would be helpful._

Can you point out where those are created, please?  I'd like to take a look at this problem, because it makes more sense to avoid creating them then delete them after the fact.  Perhaps a PR can be created for an upstream repo that makes them totally optional.

tlhumphrey2 replied:
https://github.com/hpccsystems-solutions-lab/tlh-opinionated-terraform-azurerm-hpcc/blob/add-ecl-code-security-misc/storage.tf

dcamper replied:
A couple of things:

1) If I'm reading that correctly, those storage accounts depend on:

`external_storage_config_enabled = (var.external_storage_config != null) && (var.internal_storage_enabled == false) ? true : false `

Your `hpcc.tf` file does not set either of those variables when referencing the module.

2) The link you pasted references a specific branch of that repo but your code does not.  I know that another PR request was to remove branches and it may have been this one, but the 'real' solution is to make sure the production code is on the main/master branch of the remote repo and then not reference a branch in your code.  Anyway, there may be a difference between the remote's main and specific branches, when it comes to storage account creation, and that should be explored.

The bottom line is, please try to avoid creating unneeded resources at all, rather than allowing creation and deleting after the fact.  The latter wastes time and money, and is error prone (your resource-gathering code could, logically, find something unrelated and delete it).

tlhumphrey2 replied:
internal_storage_enabled is set on line 49 of hpcc.tf
external_storage_config is set on line 69 of hpcc.tf.

dcamper replied:
I just updated my copy of that project and see that there is no main/master branch.  Furthermore, the storage.tf file within 'add-ecl-code-security-misc' branch you had cited has not been updated in quite some time.

So this is clearly a non-trivial fix.  Please spend some time investigating and, as noted above, "_...please try to avoid creating unneeded resources at all, rather than allowing creation and deleting after the fact. The latter wastes time and money, and is error prone (your resource-gathering code could, logically, find something unrelated and delete it)._"

tlhumphrey2 replied:
In locals.tf, the following code should (I thought) keep the default storages from being created:
`  admin_services_storage = local.external_storage_config_enabled ? merge(
    {
      for plane in var.external_storage_config :
      plane.category => {
        size = plane.size,
        type = plane.storage_type
      } if plane.category != "data"
  }) : var.admin_services_storage
`

tlhumphrey2 replied:
In locals.tf, there are several other places I see 'local.external_storage_config_enabled' being used.

dcamper replied:
Your last two replies are simple statements. Are there conclusions I should draw from them?  Do you have questions?  Is this issue resolved?

tlhumphrey2 replied:
I'm not sure what you want here. Did you want me to add a 'main' branch to the repo 'tlh-opinionated-terraform-azurerm-hpcc'?

dcamper replied:
Please continue to work on this issue; leaving unresolved.

dcamper commented on lite-variables.tf:
```diff
@@ -0,0 +1,218 @@
+
+variable "aks_node_sizes" {
+  description = "The VM size for each node of each node pool in the HPCC Systems. Example format is '{ roxie = \"xlarge\", serv = \"2xlarge\", spray = \"xlarge\", thor = \"xlarge\" }'.\nSee https://docs.microsoft.com/en-us/azure/virtual-machines/sizes-general for more information."
+  type = object({
```
Please be extremely careful about over-allocating resources here, as they cost the user money.  Reduce to the bare minimum if possible.  Better yet, dynamically determine the best size on a per-apply basis.

dcamper commented on lite-variables.tf:
```diff
@@ -0,0 +1,218 @@
+
+variable "aks_capacity" {
+  description = "The min and max number of nodes of each node pool in the HPCC Systems. Example format is '{ roxie_min = 1, roxie_max = 3, serv_min = 1, serv_max = 3, spray_min = 1, spray_max = 3, thor_min = 1, thor_max = 3}'."
+  type = object({
```
Please be extremely careful about over-allocating resources here, as they cost the user money.  Reduce to the bare minimum if possible.  Better yet, dynamically determine the best counts on a per-apply basis.

dcamper commented on lite-variables.tf:
```diff
@@ -53,36 +53,29 @@ variable "aks_admin_ip_cidr_map" {
 variable "aks_node_sizes" {
   description = "The VM size for each node of each node pool in the HPCC Systems. Example format is '{ roxie = \"xlarge\", serv = \"2xlarge\", spray = \"xlarge\", thor = \"xlarge\" }'.\nSee https://docs.microsoft.com/en-us/azure/virtual-machines/sizes-general for more information."
   type = object({
```
Please break this object out into scalar variables (one of the original design choices in this project), with default values for each.  Remember to update the documentation as well.

tlhumphrey2 replied:
fixed

tlhumphrey2 replied:
fixed

dcamper commented on lite-variables.tf:
```diff
@@ -0,0 +1,221 @@
+  default     = ""
+}
+
+variable "hpcc_namespace" {
```
Please break this object out into scalar variables (one of the original design choices in this project), with default values for each. Remember to update the documentation as well.

tlhumphrey2 replied:
fixed

dcamper commented on aks/locals.tf:
```diff
@@ -1,23 +1,123 @@
+    node_size         = var.aks_node_sizes.roxie
+    single_group      = false
+    min_capacity      = 1
+    max_capacity      = 3
```
This size of a Roxie pool is still hardwired.  Is that correct?

tlhumphrey2 replied:
fixed.

dcamper commented on aks/locals.tf:
```diff
@@ -1,23 +1,123 @@
+      node_size         = var.aks_node_sizes.spray
+      single_group      = false
+      min_capacity      = 3
+      max_capacity      = 6
```
Did you revert the code to hardwired sizes?

tlhumphrey2 replied:
fixed

dcamper commented on lite-locals.tf:
```diff
@@ -1,5 +1,63 @@
+output "thor_max_jobs" {
```
Is this output (and the next several lines) for debugging purposes or did you intend to have them remain for the final product?  If the latter, please update the documentation.

tlhumphrey2 replied:
These outputs are for debugging. Regarding you comments about the min_capacity and max_capacity in aks/locals.tf: all min_capacitys are 1. And, only the max_capacity for thorpool is variable because it is the only one that I can calculate.

dcamper replied:
OK.  I will keep this conversation as unresolved until the debugging statements are removed.

dcamper commented on lite-locals.tf:
```diff
@@ -8,17 +66,80 @@ locals {
   owner_name_initials = lower(join("",[for x in split(" ",local.owner.name): substr(x,0,1)]))
 
-  tags = var.extra_tags
+  /*metadata = {
```
Please delete commented-out code.

tlhumphrey2 replied:
fixed

dcamper replied:
It is not fixed.  See line 94, below, as a simple example.

dcamper commented on lite-locals.tf:
```diff
@@ -63,6 +190,36 @@ locals {
     replicas     = 6
```
Why 6?

tlhumphrey2 replied:
6 was hard coded probably my those who created the opinionated hpcc. I would think that the number of replicas probably depends on the amount of spraying that is being done. My understanding is that none of these exist until you need them (i.e. on demand). So, there is no cost until they are used.

dcamper commented on aks/locals.tf:
```diff
@@ -13,8 +13,8 @@ locals {
     single_group      = false
-    min_capacity      = var.aks_capacity.roxie_min
-    max_capacity      = var.aks_capacity.roxie_max
+    min_capacity      = 1
```
This should be zero -- or the entire node pool not built at all -- if Roxie is disabled.

tlhumphrey2 replied:
All min_capacitys are 1 now because when I set them to zero, they still showup on the portal as one.

tlhumphrey2 replied:
Fixed

dcamper commented on lite-locals.tf:
```diff
@@ -385,15 +547,16 @@ locals {
     }
     workerResources = {
-      cpu    = 4
+      cpu    = 3
```
Is 3 the best number here (I thought it was 2).

tlhumphrey2 replied:
Now, both cpu and memory for 'workerResources' are variable.

dcamper commented on lite-variables.tf:
```diff
@@ -185,6 +178,16 @@ variable "thor_num_workers" {
   default    = 2
 }
 
+variable "thor_worker_cpus" {
```
This is not something the user should be required to set.  It should be derived from the VM size you choose for the thorpool.

tlhumphrey2 replied:
Knowing the VM size, which means you know the number cpus on each node, isn't enough to determine 'thor_worker_cpus' because you can have more than one thor worker on a node.  I do have a default value for 'thor_worker_cpus' of 2. I also, restrict 'thor_worker_cpus' to 2, 4, 8, 16 because I want it to be a factor of the number of cpus on each node.

tlhumphrey2 replied:
1. Also, I now calculate 'thor_worker_ram'.

dcamper replied:
Again, this is not something the user should be required to set.

Let's simplify this for now:  Thor workers will always use 2 vCPUs.  Please remove this TF variable and use a constant of 2 cores for future computations.

tlhumphrey2 replied:
fixed

dcamper commented on lite-variables.tf:
```diff
@@ -53,36 +53,29 @@ variable "aks_admin_ip_cidr_map" {
   }
 }
+
+variable "aks_thorpool_max_capacity" {
```
The user should not be required to set the maximum number of nodes in a node pool.  The user should simply indicate the size of the thor they want (e.g. 200-way) and how many simultaneous jobs they want to run and we figure out the rest.

dcamper commented on lite-locals.tf:
```diff
@@ -1,5 +1,63 @@
+}
 locals {
-  helm_chart_timeout=600
+  ns_spec = {
```
Where is this information mapped back to Azure VM size names?

tlhumphrey2 replied:
I'm not sure what you are asking?? Are you saying that the value of 'helm_chart_timeout' should be dependent on the VM size??

dcamper replied:
I am asking about how VM sizing is now working.  It used to be that we had to choose a specific VM size and version, such as Standard_D8_v4.  Do we not need that any longer?

tlhumphrey2 replied:
In the repo: https://github.com/hpccsystems-solutions-lab/tlh-oss-terraform-azurerm-aks
here is where xlarge is translated into Standard_D4s_v4, for example: modules/node-groups/modules/node-group/local.tf

dcamper commented on hpcc/outputs.tf:
```diff
@@ -1,9 +1,21 @@
-  value       = local.hpcc_namespace
+output "eclwatch_url" {
+  description = "Print the ECL Watch URL."
+  value       = format("https://%s.%s:18010",var.a_record_name, var.aks_dns_zone_name)
```
1. During a standup we talked about using port 443 for https connectivity rather than 18010 or 18002.  Please consult your notes regarding that conversation, then modify the modules to use that port in the configuration.
2. Please do not hardcode the port number here.  Check the code, as I believe it is hardcoded elsewhere as well.

tlhumphrey2 replied:
fixed

dcamper commented on README.md:
```diff
@@ -1,18 +1,152 @@
+
+NOTE: A tutorial of this Terraform for the developer, or others who are interested, can be found [here](documentation/hpcc-tf-for-developers.md).
+
+This is a slightly-opinionated Terraform module for deploying an HPCC Systems cluster on Azure's kubernetes service (aks).  The goal is to provide a simple method for deploying a cluster from scratch, with only the most important options to consider.
```
Minor: 'kubernetes' should be capitalized.  Ensure capitalization in all external documentation.

tlhumphrey2 replied:
fixed

dcamper commented on README.md:
```diff
@@ -1,18 +1,152 @@
+
+The HPCC Systems cluster created by this module uses ephemeral storage, which is the default. This means the storage will be deleted when the cluster is deleted) But, you can also have Persistent Storage.  See the section titled [Persistent Storage](#persistent-storage), below.
+
+This repo is a fork of the excellent work performed by Godson Fortil.  The original can be found in branch, HPCC-27615 of [https://github.com/gfortil/Terraform-azurerm-hpcc].
```
Please delete this paragraph.

tlhumphrey2 replied:
fixed. I deleted the paragraph, "This repo is a fork of the excellent work performed by Godson Fortil. ....f

dcamper commented on documentation/hpcc-tf-for-developers.md:
```diff
@@ -0,0 +1,373 @@
+# For Developers: Tutorial of HPCC Easy Deploy Terraform 
+
+This tutorial explains the terraform that deploys HPCC Systems on an azure kubernetes service (aks). The terraform was designed to enable one to deploy HPCC Systems easily.
```
Minor: Ensure 'terraform', 'kubernetes', and 'github' -- when used as product names like they are here -- are capitalized in all external documentation.

tlhumphrey2 replied:
fixed

dcamper commented on documentation/hpcc-tf-for-developers.md:
```diff
@@ -0,0 +1,373 @@
+
+From the root directory of the repository one can deploy all components of the HPCC cluster. Also, one can deploy individual components of the system from these subdirectories: `vnet`, `storage`, `aks`, and `hpcc`. If you want to deploy the individual components manually, here is the order you should do the deployment: 1st `vnet`, 2nd `storage` (if you want persistent storage), 3rd `aks`, and fin ...
+
+The following sections will explain the terraform in root directory and all subdirectories.
```
Minor: The phrase 'the terraform' confused me until I realized that you really mean 'the Terraform code'.  Maybe change it throughout the document for clarity?

tlhumphrey2 replied:
Fixed

dcamper commented on documentation/hpcc-tf-for-developers.md:
```diff
@@ -0,0 +1,373 @@
+| `aks_dns_zone_resource_group_name` | string | Name of the resource group of the above dns zone. Example entry: "app-dns-prod-eastus2" |
+| `aks_enable_roxie` | boolean | Enable ROXIE? This will also expose port 8002 on the cluster. Example entry: false |
+| `aks_max_node_count` | number | The maximum number of VM nodes to allocate for the HPCC Systems node pool. Must be 2 or more. |
+| `aks_node_size` | string | The VM size for each node in the HPCC Systems node pool. Recommend "Standard_B4ms" or better. See https://docs.microsoft.com/en-us/azure/virtual-machines/sizes-general for more information. |
```
Is the description here still accurate?  I thought you were no longer using Azure VM sizes (using 'large' and 'xlarge' and such instead).

dcamper commented on documentation/hpcc-tf-for-developers.md:
```diff
@@ -0,0 +1,373 @@
+
+The subfolders, except for `scripts`, create components needed by the full system.
+
+## scripts subdirectory
```
Most of the 'here are the contents of the subdirectory' sections appear to be out of date.  Please update them.

tlhumphrey2 replied:
The statemeint, 'here are the contents of the subdirectory', no longer exists.

tlhumphrey2 replied:
Fixed

dcamper commented on documentation/hpcc-tf-for-developers.md:
```diff
@@ -0,0 +1,373 @@
+
+​     
+
+## Appendix A 
```
None of the information in these appendices adds value.  Please delete all of them.

tlhumphrey2 replied:
fixed

tlhumphrey2 commented on aks/locals.tf:
```diff
@@ -1,23 +1,123 @@
+    public_network_access_enabled = false
+    automation_account_name       = "aks-stop-demo-${random_string.name.result}"
+
+    schedule = [
```
fixed

tlhumphrey2 commented on aks/locals.tf:
```diff
@@ -45,12 +145,11 @@ locals {
-  # today        = formatdate("YYYY-MM-DD", timeadd(local.current_time, "1h"))
 
-  utc_offset = var.aks_automation.schedule[0].daylight_saving ? 4 : 5
+  utc_offset = local.aks_automation.schedule[0].daylight_saving ? 4 : 5
```
fixed

tlhumphrey2 commented on hpcc/lite-locals.tf:
```diff
@@ -0,0 +1,435 @@
+      cpu    = 2
+      memory = "4G"
+    }
+    workerResources = {
```
fixed

tlhumphrey2 commented on hpcc/lite-locals.tf:
```diff
@@ -0,0 +1,435 @@
+      query      = "3G"
+      thirdParty = "500M"
+    }
+    eclAgentResources = {
```
fixed. This was fixed earlier. But, I accidently deleted lite-locals.tf and had to do it all over. Now it is fixed.

tlhumphrey2 commented on hpcc/main.tf:
```diff
@@ -17,24 +17,33 @@ module "metadata" {
+  depends_on = [module.hpcc]
+}
+
+/*resource "null_resource" "launch_svc_url" {
```
fixed

tlhumphrey2 commented on hpcc/outputs.tf:
```diff
@@ -1,9 +1,20 @@
-  value       = local.hpcc_namespace
+output "eclwatch_url" {
+  description = "Print the ECL Watch URL."
+  value       = format("%s.%s:18010",var.a_record_name, var.aks_dns_zone_name)
```
Yes. 18010 is the correct port for https.

dcamper commented on lite.auto.tfvars.example:
```diff
@@ -0,0 +1,264 @@
+#------------------------------------------------------------------------------
+
+# The version of HPCC Systems to install.
+# Only versions in nn.nn.nn format are supported.
```
Minor: Isn't the string literal "latest" also supported?

tlhumphrey2 replied:
_Fixed_

dcamper commented on scripts/needed-auto-tfvars-files/aks/misc.auto.tfvars.example:
```diff
@@ -0,0 +1,22 @@
+  azure_kubernetes_service = {
+    tags = { "enclosed resource" = "open source aks" }
+  }
+  # azure_log_analytics_workspace = {
```
Please delete commented-out code here and elsewhere.

tlhumphrey2 replied:
Fixed

dcamper commented on scripts/needed-auto-tfvars-files/storage/storage.auto.tfvars.example:
```diff
@@ -1,8 +1,13 @@
     delete_protection         = false //Set to false to allow deletion
     prefix_name               = "adminsvc1"
     storage_type              = "azurefiles"
+    #authorized_ip_ranges = { anyone = "97.118.251.104" }
```
Please delete commented-out code here and elsewhere.

tlhumphrey2 replied:
Fixed

dcamper reviewed (changes requested):
Getting much closer.

tlhumphrey2 commented on README.md:
```diff
@@ -2,24 +2,22 @@
 * <font color="red">**helm**</font> Helm is used to deploy the HPCC Systems processes under Kubernetes.  Instructions for downloading and installing Helm are at [https://helm.sh/docs/intro/install](https://helm.sh/docs/intro/install/).
 
-* <font color="red">**kubectl**</font> The Kubernetes client (kubectl) is also required so you can inspect and manage the Azure Kubernetes cluster.  Instructions for download and installing that can be found at [https://kubernetes.io/releases/download/](https://kubernetes.io/releases/download/).  Make sure you have version 1.22.0 or later.
+* <font color="red">**kubectl**</font> The Kubernetes client (Kubectl) is also required so you can inspect and manage the Azure Kubernetes cluster.  Instructions for download and installing that can be found at [https://kubernetes.io/releases/download/](https://kubernetes.io/releases/download/).  Make sure you have version 1.22.0 or later.
```
fixed

tlhumphrey2 commented on documentation/hpcc-tf-for-developers.md:
```diff
@@ -1,11 +1,11 @@
 
-This tutorial explains the terraform that deploys HPCC Systems on an azure kubernetes service (aks). The terraform was designed to enable one to deploy HPCC Systems easily.
-The terraform can be found on github. Here is a link to it ([https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite])
+This tutorial explains the terraform code that deploys HPCC Systems on an azure Kubernetes service (aks). The terraform code was designed to enable one to deploy HPCC Systems easily.
```
fixed

tlhumphrey2 replied:
Still trying to deploy lite on my free trial azure account. They only allow 4 vCPUs. This morning I reduced max_capacity for all nodepools to 1 and replicas to 1. aks still failed with a quota error. I need at least 12 vCPUs. Next, I try to deduce the number of node pools to just 2, the system nodepool and the hpcc nodepool and see if that works.

dcamper commented on lite.auto.tfvars.example:
```diff
@@ -0,0 +1,254 @@
+# Value type: string
+# Updateable: Y
+
+hpcc_version="9.4.4"
```
Please change version to 'latest' so we don't have to keep up with versions as time goes by.

dcamper commented on lite.auto.tfvars.example:
```diff
@@ -0,0 +1,254 @@
+# Value type: number
+# Updateable: Y
+
+thor_max_jobs=2
```
Let's change the default to 1 to avoid unnecessary resource allocation for people just kicking the tires.

dcamper commented on lite.auto.tfvars.example:
```diff
@@ -0,0 +1,254 @@
+# The VM size for each roxie node in the HPCC Systems node pool.
+# Possible values are: large, xlarge, 2xlarge, and 4xlarge
+# Value type: string
+# Updateable: N
```
What does #Updateable: N mean?  Does it indicate that I should not change this value?  If that is the case, why is it a variable?

dcamper commented on lite.auto.tfvars.example:
```diff
@@ -0,0 +1,254 @@
+#------------------------------------------------------------------------------
+
+# The Azure region abbreviation in which to create these resources.
+# Must be one of ["eastus", "eastus2", "centralus"].
```
We need to remove this constraint.  We have no idea which regions our users will be using.  This list was originally for when we were deploying internal clusters for our own use and the constraint no longer applies.

dcamper commented on lite-locals.tf:
```diff
@@ -0,0 +1,525 @@
+  np1j = "${var.thor_num_workers /  local.thorWorkersPerNode }"
+  nodesPer1Job = ceil(local.np1j) == local.np1j? local.np1j : "local.nodesPer1Job, ${local.np1j}, is not an integer because var.thor_num_workers, ${var.thor_num_workers}, is not a multiple of local.thorWorkersPerNode, ${local.thorWorkersPerNode}."
+
+  thorpool_max_capacity = ceil("${ local.nodesPer1Job * var.thor_max_jobs }")
```
This does not work in all cases (one Thor worker, one max job); nodesPer1Job becomes 0.5 and fails validation.

```
│ Error: Invalid operand
│ 
│   on lite-locals.tf line 66, in locals:
│   66:   thorpool_max_capacity = ceil("${ local.nodesPer1Job * var.thor_max_jobs }")
│     ├────────────────
│     │ local.nodesPer1Job is "local.nodesPer1Job, 0.5, is not an integer because var.thor_num_workers, 1, is not a multiple of local.thorWorkersPerNode, 2."
│ 
│ Unsuitable value for left operand: a number is required.
```

dcamper commented on aks/locals.tf:
```diff
@@ -1,23 +1,120 @@
+    node_type_version = "v2"      # v1, v2
+    node_size         = var.aks_serv_node_size
+    single_group      = false
+    min_capacity      = 4
```
Please verify this min_capacity value.  Is it really needed?  I am somewhat suspicious of it, if we have trimmed the HPCC cluster down at all then we may not need 4 nodes at minimum.

dcamper commented on lite-variables.tf:
```diff
@@ -0,0 +1,224 @@
+variable "aks_azure_region" {
+  type        = string
+  description = "REQUIRED.  The Azure region abbreviation in which to create these resources.\nMust be one of [\"eastus\", \"eastus2\", \"centralus\"].\nExample entry: eastus2"
+  validation {
```
I think this is noted elsewhere, but: We should remove the region constraints.  What is there now, which was brought over from old code, pertains to our group and not to OSS.  Users should be free to place their clusters anywhere in the world.

dcamper commented on lite-variables.tf:
```diff
@@ -0,0 +1,224 @@
+}
+
+variable "aks_enable_roxie" {
+  description = "REQUIRED.  Enable ROXIE?\nThis will also expose port 8002 on the cluster.\nExample entry: false"
```
Is this port number correct?

dcamper commented on lite.auto.tfvars.example:
```diff
@@ -0,0 +1,262 @@
+#------------------------------------------------------------------------------
+
+# Enable ROXIE?
+# This will also expose port 8002 on the cluster.
```
Is 8002 correct?

dcamper reviewed (changes requested):
A few new comments, but be aware that there are still older outstanding conversations as well.

dcamper commented on lite-locals.tf:
```diff
@@ -0,0 +1,470 @@
+    }
+    workerMemory = {
+      query      = "3G"
+      thirdParty = "500M"
```
The query value is really workerResources.memory - thirdParty, rounded down.  Please change the hardcoded 3G to that computation.

The goal of this block is to divide workResources.memory into "memory for Thor" and "memory for 3rd party tools".

dcamper replied:
In your commit 81eca56d96befcdf4148fbb58a13f91ea19e429a you changed the query value to exactly the same value as workerResources.memory, which is not what was requested.  Your change runs the risk of over-allocation of RAM, leading to an Out Of Memory exception and a killed pod.

dcamper commented on README.md:
```diff
@@ -1,18 +1,151 @@
-| :---------------------------------: | :----: | :-----: |
-| `terraform-azurerm-virtual-network` |        | `VNet`  |
\ No newline at end of file
+# Deploy HPCC Systems on Azure under Kubernetes
```
I just noticed that this project creates log files in ~/tflogs, a directory that is created if it does not exist.  This directory needs to be documented so the user knows to look there for troubleshooting reasons or perhaps to clean things up after executing this code.

Bonus points for adding a mechanism that avoids that logging (or creating the log directory) as a runtime option.

dcamper replied:
Please address previous comment.

tlhumphrey2 replied:
Fixed in commit, 4d1b78f

dcamper commented on lite-locals.tf:
```diff
@@ -432,8 +433,8 @@ locals {
-      query      = format("%dG", local.thor_worker_ram)
-      thirdParty = "500M"
+      query      = format(%dG", local.workerMemory_query)
+      thirdParty = format(%dG", local.workerMemory_thirdParty)
```
This is not the same as the original.  The hardcoded value was "500M" but your computed value is "500G" -- definitely an error or OOM condition if a Thor job starts.

tlhumphrey2 commented:
In lite-locals.tf, I corrected the calculation of workMemory.query.

From: Dan S. Camper ***@***.***>
Sent: Friday, January 5, 2024 6:29 AM
To: hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite ***@***.***>
Cc: Humphrey, Timothy L (RIS-HBE) ***@***.***>; Author ***@***.***>
Subject: Re: [hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite] Main review mods 2023 12 06 (PR #55)


*** External email: use caution ***



@dcamper commented on this pull request.

________________________________

In lite-locals.tf<https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite/pull/55#discussion_r1442883531>:

> @@ -432,8 +433,8 @@ locals {

       memory = format("%dG", local.thor_worker_ram)

     }

     workerMemory = {

-      query      = format("%dG", local.thor_worker_ram)

-      thirdParty = "500M"

+      query      = format(%dG", local.workerMemory_query)

+      thirdParty = format(%dG", local.workerMemory_thirdParty)

This is not the same as the original. The hardcoded value was "500M" but your computed value is "500G" -- definitely an error or OOM condition if a Thor job starts.

-
Reply to this email directly, view it on GitHub<https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite/pull/55#pullrequestreview-1805968156>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAIUZYF7GHALK4XKQDLMAR3YM752FAVCNFSM6AAAAABAJMATMSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQMBVHE3DQMJVGY>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

________________________________
The information contained in this e-mail message is intended only for the personal and confidential use of the recipient(s) named above. This message may be an attorney-client communication and/or work product and as such is privileged and confidential. If the reader of this message is not the intended recipient or an agent responsible for delivering it to the intended recipient, you are hereby notified that you have received this document in error and that any review, dissemination, distribution, or copying of this message is strictly prohibited. If you have received this communication in error, please notify us immediately by e-mail, and delete the original message.

dcamper commented on hpcc/hpcc.tf:
```diff
@@ -1,6 +1,7 @@
 module "hpcc" {
   #source = "git@github.com:hpccsystems-solutions-lab/tlh-opinionated-terraform-azurerm-hpcc.git"
   source = "github.com/hpccsystems-solutions-lab/tlh-opinionated-terraform-azurerm-hpcc.git"
+  #source = "/home/azureuser/tlhumphrey2/RBA-terraform-azurerm-hpcc"
```
Leaving this conversation re: commented-out code as unresolved until it is completed.

tlhumphrey2 replied:
I have spoke with Godji and Milind. Godji said that he has the role of Owner and has never tried a role with fewer permissions than Owner. Milind says he has always used TFE to deploy the RBA hpcc terraform, but, believes if you use the terraform outside of TFE you would need elevated permissions.

dcamper commented on hpcc/main.tf:
```diff
@@ -30,11 +30,11 @@ module "metadata" {
 }
 
-resource "null_resource" "delete_ephemeral_storage_accounts" {
+/*resource "null_resource" "delete_ephemeral_storage_accounts" {
```
Leaving unresolved until the commented-out code is removed.

tlhumphrey2 commented:
OK. Now, I'm working on the issue of not being able to deploy into eastus2.  As far as I know there shouldn't be unresolved conversations. How to I find any unresolved conversations?

From: Dan S. Camper ***@***.***>
Sent: Monday, January 8, 2024 9:22 AM
To: hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite ***@***.***>
Cc: Humphrey, Timothy L (RIS-HBE) ***@***.***>; Author ***@***.***>
Subject: Re: [hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite] Main review mods 2023 12 06 (PR #55)


*** External email: use caution ***



@dcamper requested changes on this pull request.

No new comments, but there are still unresolved conversations. In addition, the issue of not being able to deploy into the eastus2 region is still unresolved.

-
Reply to this email directly, view it on GitHub<https://github.com/hpccsystems-solutions-lab/terraform-azurerm-hpcc-lite/pull/55#pullrequestreview-1809530653>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAIUZYEKKDGXD7UIAPQ7RVTYNQMLDAVCNFSM6AAAAABAJMATMSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQMBZGUZTANRVGM>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

________________________________
The information contained in this e-mail message is intended only for the personal and confidential use of the recipient(s) named above. This message may be an attorney-client communication and/or work product and as such is privileged and confidential. If the reader of this message is not the intended recipient or an agent responsible for delivering it to the intended recipient, you are hereby notified that you have received this document in error and that any review, dissemination, distribution, or copying of this message is strictly prohibited. If you have received this communication in error, please notify us immediately by e-mail, and delete the original message.
